### PR TITLE
Add gradient accumulation to trainers

### PR DESCRIPTION
**Pull request by fardinayar**
Hi
Recent deep-learning papers use multiple GPUs for training, which is unavailable for all users. Adding gradient accumulation to Detectron2 helps reproduce results better.

### CODE REVIEW

File: detectron2/engine/train_loop.py [hunk 1]
```diff
@@ -279,20 +285,24 @@ def run_step(self):
             losses = sum(loss_dict.values())
 
         """
-        If you need to accumulate gradients or do something similar, you can
-        wrap the optimizer with your custom `zero_grad()` method.
```
Comment (ppwwyyxx) [inline detectron2/engine/train_loop.py:283]: As the comment says, gradient accumulation should be implemented with a custom optimizer.

Comment (Erotemic) [reply detectron2/engine/train_loop.py:283]: Gradient accumulation is a pretty fundamental option. It would be nice to have it available out of the box. PytorchLightning does provide this, is there a reason detectron2 can't?

Comment (wishforgood) [reply detectron2/engine/train_loop.py:283]: Would you please show a working example? I tried to do this but failed to make it work.
